### ISSUE TITLE
[babel 7] [flow] babel crashing on react native module

### ISSUE THREAD

**Issue by spwg** [babel/babel#8379]
## Bug Report

**Current Behavior**
Crashes when run against a react native module with a `import type {...} from ...` statement. The stack trace:
```
{ SyntaxError: node_modules/react-native/Libraries/StyleSheet/StyleSheet.js: Unexpected token (18:12)

  16 | const flatten = require('flattenStyle');
  17 |
> 18 | import type {
     |             ^
  19 |   ____Styles_Internal,
  20 |   ____DangerouslyImpreciseStyle_Internal,
  21 |   ____DangerouslyImpreciseStyleProp_Internal,
    at Parser.raise (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:776:15)
    at Parser.unexpected (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:2079:16)
    at Parser.expectContextual (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:2047:41)
    at Parser.parseImport (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:5205:12)
    at Parser.parseStatementContent (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:4043:27)
    at Parser.parseStatement (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:3962:17)
    at Parser.parseBlockOrModuleBlockBody (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:4513:23)
    at Parser.parseBlockBody (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:4500:10)
    at Parser.parseTopLevel (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:3938:10)
    at Parser.parse (/Users/spencergreene/github/greyskull/node_modules/@babel/core/node_modules/babylon/lib/index.js:5304:17)
  pos: 454,
  loc: Position { line: 18, column: 12 },
  code: 'BABEL_PARSE_ERROR' }
```

**Input Code**
`npx babel node_modules/react-native/Libraries/StyleSheet/StyleSheet.js`

**Expected behavior/code**
Not to crash. To print to stdout the transformed file.

**Babel Configuration (.babelrc, package.json, cli command)**

```
> cat .babelrc
{
  "presets": ["@babel/preset-flow", "react-native"]
}
> cat package.json
{
  "name": "greyskull",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "node node_modules/react-native/local-cli/cli.js start",
    "test": "jest --no-cache",
    "build": "npx babel src/ -d lib/",
    "flow": "npx flow"
  },
  "dependencies": {
    "react-native": "0.56.0"
  },
  "devDependencies": {
    "@babel/cli": "7.0.0-beta.47",
    "@babel/core": "7.0.0-beta.47",
    "@babel/preset-flow": "7.0.0-beta.47",
    "@babel/preset-react": "7.0.0-beta.47",
    "ajv": "^6.5.2",
    "babel-core": "^7.0.0-bridge.0",
    "babel-eslint": "^8.2.6",
    "babel-jest": "^23.4.0",
    "babel-preset-flow": "^6.23.0",
    "babel-preset-react-native": "^5.0.2",
    "chai": "^4.1.2",
    "enzyme": "^3.3.0",
    "enzyme-adapter-react-16": "^1.1.1",
    "eslint": "^5.2.0",
    "flow-bin": "^0.77.0",
    "jest": "^23.4.1",
    "mocha": "^5.2.0",
    "react": "^16.4.1",
    "react-addons-test-utils": "^15.6.2",
    "react-dom": "^16.4.1",
    "react-native-mock": "^0.3.1",
    "react-test-renderer": "16.4.1",
    "regenerator-runtime": "^0.12.0"
  },
  "jest": {
    "verbose": true,
    "preset": "react-native",
    "transformIgnorePatterns": [
      "node_modules/(?!(react-native)/)"
    ]
  }
}
```

**Environment**
- Babel version(s): [v7.0.0-beta.47]
- Node/npm version: [v9.11.1, 6.2.0]
- OS: [OSX 10.13.6]
- Monorepo [yes]
- How you are using Babel: [cli]

**Possible Solution**
N/A.

**Additional context/Screenshots**
I ran into this issue while trying to setup tests with Jest for a react native project. The above stack trace is also the result of `npm test`. I've already spent a day:
- fiddling with dependencies
- googling
- reading stack overflow and github issues
- slept on this problem
- ran `rm -rf node_modules && npm install`
- made sure all `@babel/*` dependencies are on the same beta verison
- tried running the test without the `jest.transformIgnorePatterns` option set in package.json

This is my first time using Babel or Jest. Thanks for any help!

**Comment by alfredoyong** [comment]
Hello! 
Any update on this issue? I have exactly the same prob

**Comment by loganfsmyth** [comment]
The error is happening because the code in `node_modules` (rather than the user's own code), has Flowtype syntax in it. Your `.babelrc` configuration applies to your own code, but not to files in `node_modules`.

**Comment by alfredoyong** [comment]
Thanks for your insight. And what can be a solution? how to indicate that code in node_modules is using Flowtype syntax?

**Comment by alfredoyong** [comment]
Ok, found a solution that worked for me. Check Kanekotic answer: [React Native - Jest: Broken since update to 0.56. How to fix it?](https://stackoverflow.com/questions/51197405/react-native-jest-broken-since-update-to-0-56-how-to-fix-it/51225530)

**Comment by mikemorris** [comment]
**tl;dr:** Is there a recommended generic approach for transpiling React Native modules which have been published to npm untranspiled with Flow annotations in react-native@0.57 with Babel 7?

---

@loganfsmyth @hramos @kelset @ljharb It's my understanding that it has been a standard practice for React Native module authors to [gitignore `.babelrc` and publish untranspiled code to npm, potentially including Flow annotations](https://github.com/facebook/react-native/issues/10966) and rely on [metro](https://github.com/facebook/metro) for transpilation.

I'm attempting to upgrade an app from react-native@0.53 to react-native@0.57.0-rc.4 and believe the combination of a few different changes in React Native and Babel 7 have come together to make this a problem.
- [Changes to `.babelrc` directory crawling](https://github.com/babel/babel/pull/7358)
- [Separation between the React and Flow presets](https://babeljs.io/docs/en/v7-migration#separation-between-the-react-and-flow-presets)
- [Migrating from `babel-preset-react-native` to `metro-react-native-babel-preset`](https://github.com/react-native-community/react-native-releases/pull/40#issuecomment-416522617)

When including https://github.com/expo/ex-navigator (I'm quite aware it's a deprecated and unmaintained module, but I suspect other modules may exhibit similar issues), I'm seeing the following error during metro bundling:
```
error: bundling failed: SyntaxError: /Users/mikemorris/Projects/apps/packages/cava-mobile/node_modules/@expo/react-native-navigator/ExNavigator.js: Unexpected token (17:12)

  15 | import * as ExNavigatorIcons from './ExNavigatorIcons';
  16 |
> 17 | import type * as ExRoute from './ExRoute';
     |             ^
  18 |
  19 | export default class ExNavigator extends React.Component {
  20 |   static Styles = ExNavigatorStyles;
```

It appears to specifically be failing on Flow syntax, not the `import` directive. I've tried adding [`overrides`](https://babeljs.io/docs/en/options#overrides) to my `.babelrc` and creating a `babel.config.js` as suggested in https://github.com/facebook/jest/issues/6053#issuecomment-383632515, but nothing I've tried seems to get metro to pick up `@babel/preset-flow` or `@babel/plugin-transform-flow-strip-types`, and it appears the the metro config syntax changed from `rn-cli.config.js` to [a new format](https://facebook.github.io/metro/docs/en/configuration) recently.

**Comment by loganfsmyth** [comment]
@mikemorris We have some docs in http://babeljs.io/docs/en/config-files#file-relative-configuration which may help elaborate on the changes. I'd expect a `babel.config.js` to work, unless you're running the builds from a directory other than the one containing the `babel.config.js` file, but I'm also not 100% familiar with how Metro runs Babel, so it is hard to say with certainty. You should at least be able to put some `console.log`s in the config file to see if it is being loaded.

**Comment by kelset** [comment]
1. I highly recommend you don't attempt a 4 version update in one go
1. as a workaround, try adding a `babel.config.js` with your required configuration and a `.babelrc` that just writes:
```
{
  "extends": "./babel.config.js"
}
```
I think Metro still looks only for the `.babelrc` file and this should help.
But maybe it got fixed, this workaround was something I attempted during my tests with 0.56

**Comment by SMJ93** [comment]
Any update on this?

**Comment by mikemorris** [comment]
Apologies for the alarmist tone earlier.

I was able to work around this issue by forking the package and changing the `import type * as foo` syntax to import exact types instead.

The issue appears to have been a change in supported syntax migrating from an older version of Flow (_not_ an issue with Flow type stripping missing entirely as I had assumed earlier), but I wasn't able to confirm this, or when support would've changed if that was indeed the case.

I'm currently using the stock `.babelrc` file in my React Native project, with no related custom Metro config.
```
{
  "presets": ["module:metro-react-native-babel-preset"]
}
```

**Comment by AikoRamalho** [comment]
I'm facing the same issue.

Here is my .babelrc:

```javascript
{
  "presets": [
    "@babel/env", 
    "@babel/react"
  ],
  "plugins": [
    "add-module-exports", 
    "transform-object-rest-spread", 
    "@babel/transform-object-assign", 
    ["@babel/plugin-proposal-decorators", { "legacy": true }],
    "@babel/plugin-proposal-function-sent",
    "@babel/plugin-proposal-export-namespace-from",
    "@babel/plugin-proposal-numeric-separator",
    "@babel/plugin-proposal-throw-expressions",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-export-default-from"
  ],
  "env": {
    "development": {
      "plugins": [
        "@babel/transform-react-jsx-source",
      ]
    }
  }
}
```

Also, the print from the error: 
![51283549_392994461265101_830035239207174144_n](https://user-images.githubusercontent.com/15454141/52228872-974d2200-289a-11e9-8c6c-5a3a7bba1607.jpg)

**Comment by nicolo-ribaudo** [comment]
Doesn't renaming your config file work?

**Comment by nicolo-ribaudo** [comment]
Closing, since this issue is caused by a wrong configuration of Babel. `.babelrc` (or `package.json#babel`) isn't applied to node_modules.

**Comment by tednyberg** [comment]
Not sure this should be closed - I'm having the same problem despite a `babel.config.js` like the following:

```
module.exports = {
  presets: ['module:metro-react-native-babel-preset', '@babel/preset-flow'],
  plugins: [
    ['@babel/plugin-transform-flow-strip-types'],
    ['@babel/plugin-transform-typescript', {allowNamespaces: true}],
    ['@babel/plugin-proposal-decorators', {legacy: true}],
  ],
};
```
This happens in a React Native project created with `react-native init [name] --template typescript`. The Metro bundler fails with a Babel error saying **Unexpected token** whenever it comes across `import type {...} from ...` syntax in a component, such as `node_modules\react-native\Libraries\Components\Button.js`.

![image](https://user-images.githubusercontent.com/3976064/63635258-13555e80-c661-11e9-8445-8e3d0d57e32e.png)

I'm not saying my `babel.config.js` is _correct_ though - I've tried countless different setups trying to get this to work. :)
